numbat/modules/core/random.nbt

```diff
@@ -1,3 +1,100 @@
+    else random() * (a - b) + b
+
+# name: Discrete uniform distribution sampling
+# url: https://en.wikipedia.org/wiki/Discrete_uniform_distribution
+# description: Uniformly samples the integers in the interval [a, b].
+fn rand_int<T>(a: T, b: T) -> T =
+    if a <= b
+    then floor( random() * (floor(b) - ceil(a) + 1 unit_of(a)) ) + ceil(a)
+    else floor( random() * (floor(a) - ceil(b) + 1 unit_of(b)) ) + ceil(b)
+
+# name: Bernoulli distribution sampling
+# url: https://en.wikipedia.org/wiki/Bernoulli_distribution
+# description: Samples a Bernoulli random variable, that is, 1 with probability p, 0 with probability 1-p.
+#              Parameter p must be a probability (0 <= p <= 1).
+fn rand_bernoulli(p: Scalar) -> Scalar =
+    if p>=0 && p<=1
+    then (if random() < p
+        then 1
+        else 0)
+    else error("p must be a probability (0 <= p <= 1).")
+
+# name: Binomial distribution sampling
+# url: https://en.wikipedia.org/wiki/Binomial_distribution
+# description: Samples a binomial distribution by doing n Bernoulli tirals with probability p.
+#              Parameter n must be a positive integer, parameter p must be a probability (0 <= p <= 1).
+fn rand_binom(n: Scalar, p: Scalar) -> Scalar =
+    if n >= 1
+    then rand_binom(n-1, p) + rand_bernoulli(p)
+    else if n == 0
+    then 0
+    else error("n must be a positive integer.")
+
+# name: Normal distribution sampling
+# url: https://en.wikipedia.org/wiki/Normal_distribution
+# description: Samples a normal distribution with mean μ and standard deviation σ using the Box-Muller transform.
+fn rand_norm<T>(μ: T, σ: T) -> T =
+    μ + sqrt(-2 σ² × ln(random())) × sin(2π × random())
+
+# name: Geometric distribution sampling
+# url: https://en.wikipedia.org/wiki/Geometric_distribution
+# description: Samples a gemoetric distribution(the distribution of the number of Bernoulli trials with probability p needed to get one success) by inversion sampling.
+#              Parameter p must be a probability (0 <= p <= 1).
+fn rand_geom(p: Scalar) -> Scalar =
+    if p>=0 && p<=1
+    then ceil( ln(1-random()) / ln(1-p) )
+    else error("p must be a probability (0 <= p <= 1).")
+
+# A helper function for rand_poisson, counts how many samples of the standard uniform distribution need to be multiplied to deceed lim.
+fn _poisson(lim: Scalar, prod: Scalar) -> Scalar =
+    if prod > lim
+    then _poisson(lim,  prod × random()) + 1
+    else -1
+
+# name: Poisson distribution sampling
+# url: https://en.wikipedia.org/wiki/Poisson_distribution
+# description: Sampling a poisson distribution with rate λ, that is, the distribution of the number of events occurring in a fixed interval if these events occur with mean rate λ.
+#              The rate parameter λ must not be negative.
+# This implementation is based on the exponential distribution of inter-arrival times. For details see L. Devroye, Non-Uniform Random Variate Generation, p. 504, Lemma 3.3.
+fn rand_poisson<T>(λ: T) -> T =
```

Comment:
Can you give a short justification for this type signature? At first, I thought it was wrong because it should return a dimensionless *count* of events. But getting an answer like 3/min for `rand_poisson(2/min)` seems reasonable. I'm just a bit worried because 2/min = 120/hour (for example), and `rand_poisson(120/hour)` would yield different results. Yes, the sum of poisson-distributed random variables is poisson-distributed, so in that sense
```
rand_poisson(120/hour)  ~  rand_poisson(2/min) + rand_poisson(2/min) + … another 58 times
```
but it still feels a bit strange to me?

Reply:
I chose this signature because the unit of the rate conveys information about the process, it matters if the events are per minute or per hour, especially if one continues to do computations with the results.  This could also be useful when using it with different dimensions, e.g or trees/m^2.

I agree however that it is a real issue that λ is ambiguous depending on the unit, especially if the unit is not immediately clear, for example `rand_poisson(100/hour - 1/min)`. 

One way to deal with this could be to split λ up into a scalar number of events and a normalization containing units.

Probably it is safest and easiest however to convert the signature to `fn rand_poisson(λ: Scalar) -> Scalar` so it only accepts and returns scalar values like the other discrete distributions and to force the user to take care of the unit explicitly if they need it.

Reply:
> especially if the unit is not immediately clear, for example `rand_poisson(100/hour - 1/min)`.

:+1:

> Probably it is safest and easiest however to convert the signature to `fn rand_poisson(λ: Scalar) -> Scalar` so it only accepts and returns scalar values like the other discrete distributions and to force the user to take care of the unit explicitly if they need it.

It's a bit sad, but yes. I think it would be best to restrict it to scalars.


I think what might be useful is for `rand_poisson(40 trees / hectare)` to give an answer like `35 trees` (not `35 trees / hectare`), because we're asking for one particular outcome of that random experiment. So maybe the count is the more descriptive result compared to a rate, because the latter would indicate some kind of mean value. But that might be even more confusing.

And of course it's impossible to implement in this form, because there is no way of knowing what is the count and what is the period/interval.

We *could* use a two-argument function like you proposed, but instead of splitting up magnitude and unit, we could maybe split up the rate into count (40 trees) and interval (hectare). Like so:
```
fn rand_poisson_unitful<C, I>(count: C, interval: I) -> C = …
```
But that's also a bit weird because we would basically just throw away the second argument :smile:.

I think we can just go with the Scalar->Scalar version for now, unless we still come up with a good solution.

Reply:
I agree, a scalar function is the cleanest way and also the standard one people will be familiar with, unlike any two argument Poisson functions we might make up.